src/Ztm.WebApi/AddressPools/SqlReceivingAddressStorage.cs

```diff
@@ -0,0 +1,202 @@
+    {
+        readonly IMainDatabaseFactory databaseFactory;
+        readonly Network network;
+
+        public SqlReceivingAddressStorage(IMainDatabaseFactory databaseFactory, Network network)
+        {
+            if (databaseFactory == null)
+            {
+                throw new ArgumentNullException(nameof(databaseFactory));
+            }
+
+            if (network == null)
+            {
+                throw new ArgumentNullException(nameof(network));
+            }
+
+            this.databaseFactory = databaseFactory;
+            this.network = network;
+        }
+
+        public async Task<ReceivingAddress> AddAddressAsync(BitcoinAddress address, CancellationToken cancellationToken)
+        {
+            if (address == null)
+            {
+                throw new ArgumentNullException(nameof(address));
+            }
+
+            using (var db = this.databaseFactory.CreateDbContext())
+            {
+                var recAddress = await db.AddAsync<ReceivingAddressModel>(
+                    new ReceivingAddressModel
+                    {
+                        Address = address.ToString(),
+                        IsLocked = false
+                    },
+                    cancellationToken
+                );
+
+                await db.SaveChangesAsync(cancellationToken);
+
+                return ToDomain(recAddress.Entity);
+            }
+        }
+
+        public async Task<ReceivingAddress> GetAsync(Guid id, CancellationToken cancellationToken)
+        {
+            using (var db = this.databaseFactory.CreateDbContext())
+            {
+                var recv = await db.ReceivingAddresses
+                    .Include(e => e.ReceivingAddressReservations)
+                    .FirstAsync(r => r.Id == id);
+
+                return ToDomain(recv);
+            }
+        }
+
+        public Task<IEnumerable<ReceivingAddress>> ListReceivingAddressAsync(AddressFilter filter, CancellationToken cancellationToken)
+        {
+            using (var db = this.databaseFactory.CreateDbContext())
+            {
+                IQueryable<ReceivingAddressModel> query = db.ReceivingAddresses.Include(a => a.ReceivingAddressReservations);
+
+                if (filter.HasFlag(AddressFilter.Available))
+                {
+                    query = query.Where(a => !a.IsLocked);
+                }
+
+                if (filter.HasFlag(AddressFilter.NeverUsed))
+                {
+                    query = query.Where(a => !a.ReceivingAddressReservations.Any());
+                }
+
+                return Task.FromResult<IEnumerable<ReceivingAddress>>(query.Select(r => ToDomain(r)).ToList());
+            }
+        }
+
+        public async Task ReleaseAsync(Guid id, CancellationToken cancellationToken)
+        {
+            using (var db = this.databaseFactory.CreateDbContext())
+            using (var tx = db.Database.BeginTransaction())
```

Comment:
The default isolation level is quite dangerous for operations we are going to do, which usually is Read Committed. Better to use Repeatable Read instead.